### PR TITLE
Issue 3 testing

### LINKED ISSUE #3: Testing
We should add unittests to the repo, and require all environments to be covered.

For instance, we should test that `env.step(action)` and `env.reset()` retrieve elements which are in line with env.observation_space given an action from env.action_space. Also, we should test `isinstance(env.observation_space, scigym.Box)` and `isinstance(env.action_space_space, scigym.Discrete)`.

Any more ideas for tests, @R-Sweke ?

### PR DESCRIPTION
**Pull request by R-Sweke**
In order to close #3, and following openAI gym, I have added unit tests (via pytest) for scigym environments, which can be found in the folder `scigym/envs/tests/`. In particular,  I have added:

1. `test_envs.py`: This is as per openAI gym and tests basic attributes of all environments, i.e. that the observation spaces, action spaces etc are consistent. It also tests random rollouts for a specified list of specific environments. Currently only the `surfacecode-decoding-v0` environment is included in this list, but we may want to add more environments to the list as environments are added.
2. `test_determinism.py`: Again, as per openAI gym this tests that deterministic aspects of all environments are indeed deterministic. If an environment is in fact nondeterministic (in response to actions) then this should be indicated with `nondeterministic=True` when registering this environment (see registration of `surfacecode-decoding-v0` for an example) , so that it is excluded from aspects of this unit test.
3. `README.md` containing details of how to run the unit tests.

As per openAI gym, further environment specific tests can now be easily added into the directory  `scigym/envs/tests/. Two things to note:

1. openAI gym includes a test of the registration process `test_registration.py`. I have not included this unit test in scigym, as the registration of environments is handled by `gym` methods.
2. We currently have cartpole included as a test environment. I have kept it there for now, although I suggest removing this, as it is no longer really necessary.

### CODE REVIEW

**Comment by HendrikPN** [comment]
Great work! Thanks Ryan. 
I already had a quick glance at the branch and it seems alright at first sight. I just have a few preliminary statements:
1. You may already remove the cartpole environment everywhere where it appears (including the `test_envs` folder.
2. Ideally we want to avoid having to run `python -m pytest test_envs.py test_determinism.py` every time we want to perform a test. So this is where travis CI comes in (https://travis-ci.org/). So after merging this branch or while still working on it, we should include a travis implementation. This provides the advantage of automated testing as in https://github.com/openai/gym . 
3. Your `decoding` environment seems to throw a couple of `DeprecationWarnings`. Can you fix ithat, or is that unnecessary?
4. Don't forget to add to the What's New section in the README

**Comment by R-Sweke** [comment]
great - about your comments:

1. I have removed the cartpole test environment completely
2. My preference would be to merge this branch and add travis CI compatibility later - I'm not so sure that is necessary for the MVP.
3. This is tricky. The surface code env does not explicitly use the functions throwing warnings anywhere, but I think they arise because of the tensorflow/keras versions that were used to train the referee decoders (which I would prefer not to go back to). If I use tensorflow version 1.5 (which is not a stable version) as opposed to tensor flow 1.14 (which is a stable version) with keras 2.2.4 then I do not get these deprecation warnings when instantiating the environment, but I get 211 deprecation and pending deprecation warnings when running the unit tests. My suggestions would be to (a) specify the required tensorflow and keras versions explicitly (as I have done now in setup.py, and should be addressed more carefully when resolving issue #2) and (b) filter the deprecation and pending deprecation warnings from the unit tests, as discussed here: [https://docs.pytest.org/en/latest/warnings.html](https://docs.pytest.org/en/latest/warnings.html).
4. Done!

File: .gitignore [hunk 1]
```diff
@@ -22,3 +22,28 @@ logs/
 
 .ipynb_checkpoints
 ghostdriver.log
+
+#Pycharm IDE
```
Comment (HendrikPN) [inline .gitignore]: Can we not just remove `.idea`?

File: scigym/envs/tests/README.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#### Unit Tests
+
+This folder contains unit tests for scigym environments, adapted from the openAI gym suite of tests.
+
+Run these tests by exececuting the following commands from within this directory:
+
+```python
+pytest test_envs.py test_determinism.py
```
Comment (HendrikPN) [inline scigym/envs/tests/README.md]: You can also run `python -m pytest` in terminal. The -m flag lets python know that you are using a python module.

File: scigym/envs/tests/__init__.py [hunk 3]
```diff
@@ -0,0 +1 @@
+# This can remain empty...
```
Comment (HendrikPN) [inline scigym/envs/tests/__init__.py]: Generally, it is recommended to add an empty line at the end of a file (see [here](https://stackoverflow.com/questions/2287967/why-is-it-recommended-to-have-empty-line-in-the-end-of-file))

File: scigym/envs/tests/test_determinism.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            v_a = a[k]
+            v_b = b[k]
+            assert_equals(v_a, v_b)
+    elif isinstance(a, np.ndarray):
+        np.testing.assert_array_equal(a, b)
+    elif isinstance(a, tuple):
+        for elem_from_a, elem_from_b in zip(a, b):
+            assert_equals(elem_from_a, elem_from_b)
+    else:
+        assert a == b
```
Comment (HendrikPN) [inline scigym/envs/tests/test_determinism.py]: new line

File: scigym/envs/tests/test_envs.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    for env in [envs.make('surfacecode-decoding-v0')]:
+        agent = lambda ob: env.action_space.sample()
+        ob = env.reset()
+        for _ in range(10):
+            assert env.observation_space.contains(ob)
+            a = agent(ob)
+            assert env.action_space.contains(a)
+            (ob, _reward, done, _info) = env.step(a)
+            if done: break
+        env.close()
```
Comment (HendrikPN) [inline scigym/envs/tests/test_envs.py]: new line

File: setup.py [hunk 6]
```diff
@@ -7,8 +7,7 @@
 
 # Environment-specific dependencies.
 extras = {
-    'test_envs': ['PyOpenGL'],
-    'surfacecode': ['tensorflow', 'keras']
+    'surfacecode': ['tensorflow==1.5', 'keras==2.2.4']
```
Comment (HendrikPN) [inline setup.py:10]: Nice! That makes it better :)

File: setup.py [hunk 7]
```diff
@@ -27,7 +26,7 @@
       packages=[package for package in find_packages()
                 if package.startswith('scigym')],
       zip_safe=False,
-      install_requires=['gym==0.11.0',],
+      install_requires=['gym==0.11.0','pytest'],
```
Comment (HendrikPN) [inline setup.py]: Can you remove `pytest` and add it as `tests_require=['pytest']` below?
Also, I think we want to use `'gym==0.14.0'`. See my general comments to fix the arising issues.

File: scigym/envs/tests/test_envs.py (hunk 5 above)
Comment (HendrikPN) [inline scigym/envs/tests/test_envs.py]: new line

File: scigym/envs/tests/test_determinism.py (hunk 4 above)
Comment (HendrikPN) [inline scigym/envs/tests/test_determinism.py]: new line

File: scigym/envs/tests/__init__.py (hunk 3 above)
Comment (HendrikPN) [inline scigym/envs/tests/__init__.py]: Generally, it is recommended to add an empty line at the end of a file (see [here](https://stackoverflow.com/questions/2287967/why-is-it-recommended-to-have-empty-line-in-the-end-of-file))

File: scigym/envs/tests/README.md (hunk 2 above)
Comment (HendrikPN) [inline scigym/envs/tests/README.md]: You can also run `python -m pytest` in terminal. The -m flag lets python know that you are using a python module.

File: .gitignore (hunk 1 above)
Comment (HendrikPN) [inline .gitignore:26]: Can we not just remove `.idea`?

**Review by HendrikPN — CHANGES_REQUESTED** [review]
Looks good! I don't really have any general comments, I left a few specific comments in the files.

Comments:
+ Can we also remove the `test_envs` _folder_?
+ I think we might be using an outdated version of `gym`. We might need the newest version, i.e. 0.14.0. I tried changing the gym version and the tests throw errors:
i. Can you remove `from gym.utils import reraise` from the `__init__` as [here](https://github.com/openai/gym/commit/d11196a896d20ebcd5e4e7a4aa6f3bc7ac96d1b1). 
ii. Then, to remove the last error, you need to change [this line](https://github.com/HendrikPN/scigym/blob/issue_3_testing/scigym/envs/tests/spec_list.py#L9) to 
```
spec_list = [spec for spec in sorted(envs.registry.all(), key=lambda x: x.id) if spec.entry_point[:6] == "scigym" and not should_skip_env_spec_for_tests(spec)]
```

Please double check that this resolves all version issues for testing.

File: setup.py (hunk 7 above)
Comment (R-Sweke) [reply setup.py]: Done. When I run the tests now I get 17 warnings, all of which are deprecation warnings (which are harmless because of the fact that the Keras version has been specified). 

It is possible to filter these warnings so that they don't appear, as described here:

https://docs.pytest.org/en/latest/warnings.html

I'm not sure whether it is better to leave the warnings explicitly (as apparently suggested by PEP-0565) or to filter them out to be neater. I think I'm leaning towards leaving the warnings there, in case other deprecation warnings come up in the future which we do need to address.

File: scigym/envs/tests/test_envs.py (hunk 5 above)
Comment (R-Sweke) [reply scigym/envs/tests/test_envs.py]: done.

File: scigym/envs/tests/test_determinism.py (hunk 4 above)
Comment (R-Sweke) [reply scigym/envs/tests/test_determinism.py]: done.

File: scigym/envs/tests/__init__.py (hunk 3 above)
Comment (R-Sweke) [reply scigym/envs/tests/__init__.py]: done.

File: scigym/envs/tests/README.md (hunk 2 above)
Comment (R-Sweke) [reply scigym/envs/tests/README.md]: Changed - this is definitely nicer.

File: .gitignore (hunk 1 above)
Comment (R-Sweke) [reply .gitignore:26]: At the moment I am using the .gitignore suggested by PyCharms - as this is a popular editor I would suggest leaving it as it is?

**Review by HendrikPN — APPROVED** [review]
Looks all good to me! Thanks Ryan!
